### PR TITLE
Prevent getFunctionParent from returning Program

### PR DESCRIPTION
**Pull request by sarupbanskota**
Works towards #5828

| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | 
| Major: Breaking Change?  | `getFunctionParent` will no longer return `Program`
| Minor: New Feature?      | 
| Deprecations?            | 
| Spec Compliancy?         | 
| Tests Added/Pass?        | 
| Fixed Tickets            | #5828 
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       |

### CODE REVIEW

File: packages/babel-traverse/src/path/ancestry.js [hunk 1]
```diff
@@ -33,7 +33,11 @@ export function find(callback): ?NodePath {
  */
 
 export function getFunctionParent(): ?NodePath {
-  return this.findParent(path => path.isFunction() || path.isProgram());
+  const parent = this.findParent(path => path.isFunction() || path.isProgram());
+  if (parent.isProgram()) {
```
Comment (jridgewell) [inline packages/babel-traverse/src/path/ancestry.js]: Could just check if it's `null` and remove the test in callback.

Comment (sarupbanskota) [inline packages/babel-traverse/src/path/ancestry.js]: @jridgewell umm didn't follow.. 🤔 

You meant just to do something like: `return parent.isFunction() ? parent : null;`?

I added the `isProgram` check explicitly so I could throw an error...

Comment (jridgewell) [inline packages/babel-traverse/src/path/ancestry.js]:
```js
const parent = this.findParent(p => p.isFunction());
if (!parent) {
  throw new Error("getFunctionParent won't find Program anymore");
}
return parent;

Comment (sarupbanskota) [inline packages/babel-traverse/src/path/ancestry.js]: ah! ok, did just that :)

**Comment by sarupbanskota** [comment]
I expect lots of plugin tests to fail, but otherwise, to solve the original issue itself, do you think I got it right so far @jridgewell @boopathi?

File: packages/babel-traverse/src/scope/index.js [hunk 2]
```diff
@@ -830,7 +832,7 @@ export default class Scope {
         return scope;
       }
     } while ((scope = scope.parent));
-    throw new Error("We couldn't find a Function or Program...");
+    return null;
```
Comment (boopathi) [inline packages/babel-traverse/src/scope/index.js]: why is this change required ?

File: packages/babel-traverse/src/path/ancestry.js [hunk 3]
```diff
@@ -33,7 +34,11 @@ export function find(callback): ?NodePath {
  */
 
 export function getFunctionParent(): ?NodePath {
-  return this.findParent(path => path.isFunction() || path.isProgram());
+  const parent = this.findParent(p => p.isFunction());
+  if (!parent) {
+    throw new Error("getFunctionParent won't find Program anymore");
+  }
```
Comment (boopathi) [inline packages/babel-traverse/src/path/ancestry.js]: I think we should print warning and not throw an error.

File: packages/babel-traverse/src/scope/index.js [hunk 4]
```diff
@@ -90,7 +90,9 @@ const collectorVisitor = {
     //if (path.isFlow()) return;
 
     // we've ran into a declaration!
-    path.scope.getFunctionParent().registerDeclaration(path);
+    const parent =
+      path.scope.getFunctionParent() || path.scope.getProgramParent();
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/src/scope/index.js]: This wouldn't work, because `getFunctionParent()` either returns the function path or throws. It never returns a falsy value and thus `getProgramParent()` is never called.

Comment (sarupbanskota) [inline packages/babel-traverse/src/scope/index.js]: @nicolo-ribaudo you're correct, that's why I made it to return `null` if it can't find a function path - https://github.com/babel/babel/pull/5923#discussion_r127225252

File: packages/babel-traverse/src/scope/index.js (hunk 2 above)
Comment (sarupbanskota) [inline packages/babel-traverse/src/scope/index.js]: because in another instance of the code, it freaks out https://github.com/babel/babel/pull/5923#discussion_r127225960

maybe there's a better way to deal with it though, open to suggestions :)

File: packages/babel-traverse/src/scope/index.js (hunk 4 above)
Comment (nicolo-ribaudo) [inline packages/babel-traverse/src/scope/index.js]: Oh sorry, I confused it with path.getFunctionParent() 😅

File: lib/types.js [hunk 5]
```diff
@@ -857,7 +857,7 @@ type BabelNodeExpressionWrapper = BabelNodeExpressionStatement | BabelNodeTypeCa
 type BabelNodeFor = BabelNodeForInStatement | BabelNodeForStatement | BabelNodeForOfStatement;
 type BabelNodeForXStatement = BabelNodeForInStatement | BabelNodeForOfStatement;
 type BabelNodeFunction = BabelNodeFunctionDeclaration | BabelNodeFunctionExpression | BabelNodeObjectMethod | BabelNodeArrowFunctionExpression | BabelNodeClassMethod;
-type BabelNodeFunctionParent = BabelNodeFunctionDeclaration | BabelNodeFunctionExpression | BabelNodeProgram | BabelNodeObjectMethod | BabelNodeArrowFunctionExpression | BabelNodeClassMethod;
+type BabelNodeFunctionParent = BabelNodeFunctionDeclaration | BabelNodeFunctionExpression | BabelNodeObjectMethod | BabelNodeArrowFunctionExpression | BabelNodeClassMethod;
```
Comment (sarupbanskota) [inline lib/types.js]: yeah, I'm hoping that was the right thing to do ..

File: packages/babel-traverse/src/path/ancestry.js (hunk 3 above)
Comment (boopathi) [inline packages/babel-traverse/src/path/ancestry.js]: This should just be 

```js
return this.findParent(path => path.isFunction());
```

File: packages/babel-traverse/src/scope/index.js (hunk 2 above)
Comment (boopathi) [inline packages/babel-traverse/src/scope/index.js]: Scope's `getFunctionParent` throws. paths' `getFunctionParent` doesn't throw and silently fails by returning `null`. Looks like you have swapped this behaviour.

Comment (sarupbanskota) [inline packages/babel-traverse/src/scope/index.js]: to keep both consistent, what do you think of not throwing in both cases and returning `null`? 
(although i'm letting scope's `getProgramParent()` throw since that would mean something is weird?)

Comment (loganfsmyth) [inline packages/babel-traverse/src/scope/index.js]: I don't have strong feelings on this decision, but I do agree that the name and the behavior seem contradictory.

If we do go with this change, the doc comment on this function also needs to be updated.

File: packages/babel-traverse/src/path/ancestry.js [hunk 6]
```diff
@@ -33,7 +34,7 @@ export function find(callback): ?NodePath {
  */
 
 export function getFunctionParent(): ?NodePath {
-  return this.findParent(path => path.isFunction() || path.isProgram());
```
Comment (hzoo) [inline packages/babel-traverse/src/path/ancestry.js]: This should be documented as a breaking change too
